filter-process: Ignore file size in working copy when cleaning

cynix opened the pull request:
Sometimes when Git sends us the clean command, the target file in the
working copy does not have the correct size (e.g. it's empty). Since
Git will feed us the blob to clean via stdin, we have to read it all
anyway, especially if we're invoked via filter-process where leaving
any data unread will cause our Pktline's reader to not be positioned
correctly for parsing the next command. Don't trust the file size on
disk when determining how much to copy.

Resolves #4974 (specifically, the secondary problem reported in that issue).

cynix committed:
Add test case for #4974

cynix committed:
filter-process: Ignore file size in working copy when cleaning.

Sometimes when Git sends us the clean command, the target file in the
working copy does not have the correct size (e.g. it's empty). Since
Git will feed us the blob to clean via stdin, we have to read it all
anyway, especially if we're invoked via filter-process where leaving
any data unread will cause our Pktline's reader to not be positioned
correctly for parsing the next command. Don't trust the file size on
disk when determining how much to copy.

cynix committed:
Improove comment in test util

cynix commented on t/cmd/lfstest-badlocalfile.go:
```diff
@@ -0,0 +1,72 @@
+	//
+	// If we simply truncate the file to 0 bytes in situ, Git
+	// may not be able to finish reading its full contents, and
+	// might send us fewer packets than actually needed.
```
I don't know how to actually reproduce what Git is doing when it invokes our filter-process with incorrect files in the working copy, so I'm using this util to simulate it. Let me know if you figure out a way to reproduce this natively.

chrisd8088 replied:
I think this is a perfectly reasonable way to simulate the problem!  Most likely some other process is involved which removes or alters the working copy file after Git opens it but before Git LFS can check its size, so this is a clever and effective way to simulate the issue for our test suite.

chrisd8088 commented:
Hey, thanks very much for this PR and for the investigation into the problems reported #4974, which have proven challenging to replicate!  We really appreciate this kind of contribution from our community of users.

I think this PR deserves a thoughtful review and so I hope to devote a solid chunk of time to it as soon as I can.  My apologies in advance if that takes a week or two; I can assure you it's not because we're not interested in this potential fix for a longstanding issue!

cynix committed:
Fix formatting

chrisd8088 committed:
Merge branch 'main' into fix_clean

chrisd8088 commented:
Thank you again for this PR and for your patience, @cynix!

I've started working on a review, and will get that done as soon as possible so you can see my suggested revisions, which are for the most part quite minor, at least as regards the work in this PR at the moment.

However, there are some ancillary changes I think we'll need to make as a consequence of this PR, and so I may try to implement those in another PR which we could merge ahead of this one, or ask if I can contribute some commits to this PR (for which I'd need an invitation to collaborate on your repository).

My apologies for the delay in replying, and thanks again very much for this bug fix!

cynix commented:
@chrisd8088 Thanks for the update. I've invited you as a collaborator, feel free to add commits as you see fit. I'm guessing you're thinking of finding a way to still report progress to the user even though we can't trust the file size? (e.g. use the file size on disk as a hint for progress display only, not for determining how much to copy, or something like that) :D

chrisd8088 commented:
> I'm guessing you're thinking of finding a way to still report progress to the user even though we can't trust the file size?

Yes, that's one of the issues, indeed.  At the moment, the file size in the progress log (when enabled with `GIT_LFS_PROGRESS`) will have `-1` as the denominator in the "number of bytes written so far" field, so I worked up some changes to provide an alternative (along with a bunch of tests), but then noticed a more subtle issue, which is that when `total` is `-1`, we inadvertently defeat the 200ms [throttling](https://github.com/git-lfs/git-lfs/blob/97f47b9cd2102273287066e0225a6c10457c401a/lfs/util.go#L58) of writes to the progress log, so we write to the log on every callback.

Since noticing that I've been pulled off in another direction, but will return to this topic shortly, I very much hope!

chrisd8088 commented:
Hey, I just wanted to note that although I haven't had time to circle back to this PR and the outstanding progress log issues, I will do so in time because I think this is a valuable change and I'd like to get it merged.

Thanks again for your contribution and your patience!

chrisd8088 commented:
It's taken me much longer than I expected to circle back to this PR, but I have not forgotten about it!

For my own reference, I wanted to note that the progress log (when enabled with `GIT_LFS_PROGRESS`) will already have a `-1` in the denominator of the bytes-received field if a server replies to an object transfer request with a `Content-Encoding: gzip` header and a gzipped response body.

The Go HTTP client automatically inserts an `Accept-Encoding: gzip` request header, as described in #6193, so a server is within its rights to reply with a gzipped response.  When the `net/http` package from the Go standard library has inserted the `Accept-Encoding: gzip` request header on its own, and the server returns a gzipped response with a `Content-Encoding: gzip` header, the `ContentLength` value of the `Response` structure is [set](https://github.com/golang/go/blob/9777ceceec8fee294d038182739cab7c845ad2d1/src/net/http/transport.go#L2461) to `-1`, and we then just pass that on to our progress meter.

Under these conditions, we'll also bypass our throttling of progress log messages, as I noted above in https://github.com/git-lfs/git-lfs/pull/6011#issuecomment-2761801198.

Enabling a progress log with `GIT_LFS_PROGRESS="$TRASHDIR/progress.log"` in our `storage upload with compression` [test](https://github.com/git-lfs/git-lfs/blob/bb65882304a655ffa8abf2be6922e53ff18af5a5/t/t-push.sh#L958-L998) is sufficient to demonstrate this behaviour.

cynix committed:
Merge remote-tracking branch 'origin/main' into fix_clean

chrisd8088 committed:
t/cmd/lfstest-badlocalfile.go: avoid truncation

In a prior commit in this PR we introduced a new test helper utility
named "lfstest-badlocalfile", which pipes Git packet-line data from
its standard input to its standard output, and when a file name is
detected in the packet data, the utility replaces that file with
an empty file.

This helper program allows the "filter process: don't trust file
size on disk when cleaning" test in our "t/t-filter-process.sh" shell
script to reproduce the situation where a file's data is piped to the
"git lfs filter-process" command but the file's size on disk is
altered before the command processes the data.  We added this
test in the same prior commit in which we also introduced the
"lfstest-badlocalfile" utility.

After our new helper program detects a file name in the incoming Git
packet data, if Git has also specified that a "clean" filter
operation is in progress, the utility first renames the file at
that location with a ".orig" file extension.  The program then
attempts to create a file at the original location and truncates
its size to zero.

We can simplify this second step slightly by using the O_EXCL flag
from the Go standard library's "os" package to force the OpenFile()
function from the same package to fail unless it creates a new file
rather than opening an existing file.  Since a new file must by
definition be empty, we then no longer need to explicitly try to
truncate the file to zero bytes.

Note that in the context of our "filter process: don't trust file
size on disk when cleaning" test we expect our utility to always
be able to create a new file, so if this is not the case, we just
report a panic condition and halt the program immediately.

chrisd8088 committed:
commands: ignore file size when cleaning in merges

In a prior commit in this PR we revised the clean() function in our
"commands/command_clean.go" source file so that when it is passed
a non-empty "fileName" parameter, the function no longer tries to
determine the size of the file at the location given by that
parameter.

Previously, the function would invoke the Stat() function from the
Go standard library's "os" package to attempt to read the size of
the given file.  If that succeeded, and if the clean() function's
"fileSize" parameter was negative, the file size returned by the
Stat() function would be used in the subsequent call to the Clean()
method of the GitFilter structure in our "lfs" package.

Instead, the clean() function now just passes its "fileSize"
parameter directly to the Clean() method, without modification.
This change ensures that we do not pass an incorrect size to the
Clean() method, as might previously occur under the circumstances
reported in the secondary issue from #4974.

Specifically, as described in the following comment, when Git streamed
a file's data to our "git lfs filter-process" command, if the
corresponding file on disk was altered before the command checked
its size, an error would be reported:

  https://github.com/git-lfs/git-lfs/issues/4974#issuecomment-2717907543

Both the "git lfs filter-process" command and the "git lfs clean"
command pass negative values for the "fileSize" parameter when they
invoke the clean() function, since they do not receive file size
information from Git during "clean" filter operations.

By contrast, the "git lfs migrate import" command passes non-negative
values for the "fileSize" parameter when it calls the clean()
function, because it is processing Git blobs from existing commits
in a repository's Git history, and therefore has direct access to
the sizes of the individual blobs.

One other Git LFS command also invokes the clean() function, the
"git lfs merge-driver" command.  When configured by a user, this
command may be called by Git during merge operations, and will be
passed the locations of several temporary files created by Git,
each of which represents a different revision of a file.  If the
command can automatically resolve any conflicts between these
revisions, it is expected to write its output back to a specific
temporary file.

The "git lfs merge-driver" command expects to handle merge conflicts
between Git LFS pointers, which Git cannot resolve using its own
merge machinery.  Thus if the command can resolve a conflict, it
must write a valid Git LFS pointer into the specified output file,
which in turn means the command must invoke the clean() function
to create the pointer data.  When it runs this function, it passes
a file descriptor to the output file as the "to" parameter, and
another file descriptor for a different, additional temporary file
as the "from" parameter.  This additional temporary file is not
one of the files created by Git, but is created internally by the
"git lfs merge-driver" command to hold the actual Git LFS object
data for which the command should generate a new pointer.

At present, the command opens a file descriptor for this internal
object data temporary file and passes it as the "from" parameter
to the clean() command, and also passes its path in the "fileName"
parameter.  As well, the command runs the Stat() function from the
"os" package on the file before opening it, in order to try to
determine the file's size, which is then passed to the clean()
function in its "fileSize" parameter.

In the context of a merge driver program, however, there is no
particular need to perform the Stat() call or determine the file's
size for the clean() function.  In fact, given that the size of
the file might hypothetically change between the time of the Stat()
call and the time the file is opened, it is actually preferable for
us not to pass a potentially incorrect file size to the clean()
function.

We can therefore simplify our "git lfs merge-driver" command by
removing the Stat() call entirely and just passing a negative value
for the "fileSize" parameter of the clean() function.  This will
still result in the file being read completely, because if the
"fileSize" value is negative, the copyToTemp() method of the
GitFilter structure in our "lfs" package will continue to read
from the "from" file descriptor until EOF is reached.

We also take the opportunity to update the code comments which precede
the clean() function in our "commands" package, in order to clarify
that when the function's "fileSize" parameter is negative, this merely
indicates that the length of input object data is unknown.

chrisd8088 commented:
Believe it or not, I've been able to spend some time focusing on the issues raised by this PR, with the intent that we might include this PR in our upcoming (and long-delayed) v3.8.0 release.

PR #6285 is the first of two separate PRs I'd like to merge ahead of this PR.  The next PR will refactor our progress log callback mechanisms and expand the test suite for those.

Then I'll have just a handful of revisions for this PR, mostly to deal with a few specific details which I'll note in a final review.  I can push up the changes myself, as I've already written the necessary commits.

Thanks again for the diagnosis and fix for this bug, and for your considerable patience!

cynix commented:
Thanks for the update, very happy to see progress being made, can’t wait for your new release :)

chrisd8088 commented:
PR #6288 is the second of two PRs I'd like to merge ahead of this one, as it resolves several problems with the progress log that would be much more apparent once we no longer try to guess the size of objects during `clean` operations by "peeking" at the corresponding size of the files in the working tree.

After that I just have a small set of tweaks for this PR, most importantly removing the [block](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/lfs/gitfilter_clean.go#L73-L75) which would otherwise disable the progress log entirely during `clean` operations.  I'll write up those commits and push them into this PR's branch once we've got PR #6288 in place.

chrisd8088 committed:
t/t-filter-process.sh: use getlimit in new test

In a prior commit in this PR we revised our "clean" Git filter
commands to address the secondary issue described in #4974, and
then in a subsequent commit we introduced a new test which
validates the changes we made to our filter commands.

Specifically, the "filter process: don't trust file size on disk
when cleaning" test in our "t/t-filter-process.sh" shell script
creates the conditions described in the following comment, and then
verifies that they no longer cause the "git lfs filter-process"
command to report an error:

  https://github.com/git-lfs/git-lfs/issues/4974#issuecomment-2717907543

To reproduce the conditions described in the comment, the test
starts by creating a file larger than the size of the buffer we use
to detect and decode Git LFS pointers.  The test then adds the file
as a Git LFS object, causing its contents to be streamed to our
"clean" filter process.  However, the test also ensures that a file
smaller than the size of our pointer detection buffer is written to
the same location as the original file before our filter process
actually executes.

Since we first developed this test for this PR, in commit
9fe97b0a597e177a521ba8b1729a5dfbcd10cbcb of PR #6285 we introduced a
test utility program named "lfstest-getlimit" which can return the
value of the BlobSizeCutoff constant from our "lfs" package.  This
constant defines the size of the read buffer used to decode pointers
in the Git LFS client.

As a result, in commit ac2258e8ef441f25071c1407a7ad315ff4ef6233 of the
same PR #6285 we were able to rewrite a number of tests in our shell
test suite that previously relied on hard-coded values to create test
data of a sufficient size to exceed the Git LFS client's pointer
read buffer.

We can therefore now make the same change to our new "filter process:
don't trust file size on disk when cleaning" test, and replace the
use of a file created using a hard-coded size with one which is
dynamically sized to exceed the value returned by the "lfstest-getlimit"
when it is passed the --max-pointer-size option.  Note that the
present value of the BlobSizeCutoff constant is 1024, so our test
now just creates a file twice that size, which still guarantees it
will exceed the size of the internal pointer read buffer.

As well, we also update our new test so that it uses the head(1)
command to read a specific number of zero bytes from the "/dev/zero"
pseudo-device instead of using the dd(1) command, as we did
previously.  This change mirrors one we made to many of our other
similar tests in commit ac2258e8ef441f25071c1407a7ad315ff4ef6233
of PR #6285, where we adopted the consistent use of the "head"
command to create files of specific sizes because the "dd" command
requires several somewhat obscure options, while the "head" command
only requires a single byte-count argument.

chrisd8088 committed:
t/t-filter-process.sh: update and rename new test

In a prior commit in this PR we revised our "clean" Git filter
commands to address the secondary issue described in #4974, and
then in a subsequent commit we introduced a new test which
validates the changes we made to our filter commands.

Specifically, the "filter process: don't trust file size on disk
when cleaning" test in our "t/t-filter-process.sh" shell script
creates the conditions described in the following comment, and then
verifies that they no longer cause the "git lfs filter-process"
command to report an error:

  https://github.com/git-lfs/git-lfs/issues/4974#issuecomment-2717907543

To reproduce the conditions described in the comment, the test
starts by creating a file larger than the size of the buffer we use
to detect and decode Git LFS pointers.  The test then adds the file
as a Git LFS object, causing its contents to be streamed to our
"clean" filter process.  However, the test also ensures that a file
smaller than the size of our pointer detection buffer is written to
the same location as the original file before our filter process
actually executes.

Since we first developed this test for this PR, in commit
6cc587c2a5eca3a6562763268533a6d5b9c65f59 of PR #6285 we revised a
number of the pre-existing tests in our "t/t-filter-process.sh" shell
script to better document their intent and to bring them into closer
alignment with each other.

We therefore now make several adjustments to our new "filter process:
don't trust file size on disk when cleaning" test so that it conforms
with the current implementation of the other tests in the same script.

We also expand the script to perform one additional check, and take
the opportunity to slightly rename the test to help clarify that it
expects the Git LFS client to entirely ignore any file in the working
tree which might (or might not) correspond to the source of the data
Git streams to the "git lfs filter-process" command.

First, we rename the test's "oid" shell variable to "contents_oid",
to match the naming scheme found in our other tests.

Next, we add a comment to explain how the "lfstest-badlocalfile"
helper utility that we introduced in a prior commit in this PR is
used by the test to replicate the original issue (i.e., the
secondary issue reported in #4974).

As well, we capture the output of the "git add" command into a log
file and check that this file does not contain the error message
"unknown command", which was a symptom of the original problem.
Because our use of the tee(1) command to create the log file obscures
the exit code from the "git add" command, we also explicitly check
that command's exit code, following the idiom established by many
of our other shell tests.

Finally, we rename the test to "filter process: ignore file size on
disk when cleaning", and change the name of the repository the test
creates to match the new test name.

chrisd8088 committed:
t/t-clean.sh: add file size mismatch test

In a prior commit in this PR we revised our "clean" Git filter
commands to address the secondary issue described in #4974, and
then in subsequent commits we introduced and refined a new test
which validates the changes we made to our filter commands.

Specifically, the "filter process: ignore file size on disk when
cleaning" test in our "t/t-filter-process.sh" shell script creates
the conditions described in the following comment, and then verifies
that they no longer cause the "git lfs filter-process" command to
report an error:

  https://github.com/git-lfs/git-lfs/issues/4974#issuecomment-2717907543

(Note that when we first added our new test, we initially assigned
it the name "filter process: don't trust file size on disk when
cleaning", which we then changed slightly in a subsequent commit
in this PR.)

To reproduce the conditions described in the comment on issue #4974,
the test starts by creating a file larger than the size of the
buffer we use to detect and decode Git LFS pointers.  The test then
adds the file as a Git LFS object, causing its contents to be
streamed to our "clean" filter process.  However, the test also
ensures that a file smaller than the size of our pointer detection
buffer is written to the same location as the original file before
our filter process actually executes.

Although we have two "clean" filter commands, the test only exercises
the "git lfs filter-process" command, which since Git v2.11.0 is the
preferred mechanism by which our "clean" filter is applied to
Git LFS files.

Users with an older version of Git or a non-standard configuration may
still utilize our "git lfs clean" command for the same purpose,
although it can only process a single file.  By contrast, the
"git lfs filter-process" command supports Git's long-running filter
protocol and can process multiple files asynchronously.

As the problem reported in the secondary issue from #4974 applies
to both of the commands that implement our "clean" filter, we now
add a new test to our "t/t-clean.sh" script which reproduces the
same conditions as the "filter process: ignore file size on disk
when cleaning" test, but exercises the "git lfs clean" command
instead of the "git lfs filter-process" command.

One convenient simplification which is possible with this test is
that we do not require the use of the "lfstest-badlocalfile" test
helper utility to simulate the necessary conditions, as we do in
the corresponding test of the "git lfs filter-process" command.
This utility pipes Git packet-line data from its standard input to
its standard output, and when a file name is detected in the packet
data, the utility replaces that file with an empty file.  That allows
our test of the "git lfs filter-process" command to reproduce the
situation where a file's data is piped to the command but the file's
size on disk is altered before the command processes the data.

However, in the case of the "git lfs clean" command, since it
only processes a single file at a time, we can simply pipe the
data from another file to the command after first creating an
empty file whose name is passed to the command as its sole
argument.

Note that without the revisions we made to our "clean" Git filter
commands in earlier commits in this PR, the conditions replicated
by our new "clean: ignore file size on disk when cleaning" test
would cause the "git lfs clean" command to return an incorrect
Git LFS pointer, which would in turn cause the test to fail.

This would occur because the pointer's "size" field would be set
to 1024 bytes, as the command would only consume enough of its
input data to fill the pointer read buffer, whose length is set
by the BlobSizeCutoff constant from our "lfs" package.  Since the
input data's length would be incorrectly determined to be zero (due
to the empty file), the copyToTemp() method of the GitFilter
structure in our "lfs" package would skip reading any additional
input data, assuming it to be complete.  As a consequence, the
command would output a Git LFS pointer for a file of 1024 bytes,
regardless of the fact that more input data remained to be read.

chrisd8088 committed:
lfs,t: restore progress log in clean filter

In a prior commit in this PR we revised the clean() function in our
"commands/command_clean.go" source file so that when it is passed
a non-empty "fileName" parameter, the function no longer tries to
determine the size of the file at the location given by that
parameter.

Previously, the function would invoke the Stat() function from the
Go standard library's "os" package to attempt to read the size of
the given file.  If that succeeded, and if the clean() function's
"fileSize" parameter was negative, the file size returned by the
Stat() function would be used in the subsequent call to the Clean()
method of the GitFilter structure in our "lfs" package.

Instead, the clean() function now just passes its "fileSize"
parameter directly to the Clean() method, without modification.
This change ensures that we do not pass an incorrect size to the
Clean() method, as might previously occur under the circumstances
reported in the secondary issue from #4974.

Specifically, as described in the following comment, when Git streamed
a file's data to our "git lfs filter-process" command, if the
corresponding file on disk was altered before the command checked
its size, an error would be reported:

  https://github.com/git-lfs/git-lfs/issues/4974#issuecomment-2717907543

Both the "git lfs filter-process" command and the "git lfs clean"
command pass negative values for the "fileSize" parameter when they
invoke the clean() function, since they do not receive file size
information from Git during "clean" filter operations.  Hence the
Clean() method now always receives a negative value in its "fileSize"
parameter during "clean" filter operations.

In the normal case, when no Git LFS pointer extension programs are
configured, the Clean() method invokes the GitFilter structure's
copyToTemp() method to read the object data piped from Git, check
if it already contains a valid Git LFS pointer, and otherwise
calculate the data's SHA-256 hash while spooling the data to a
temporary file.  To write the object data to the file, the
copyToTemp() method uses the CopyWithCallback() function from our
"tools" package.

The purpose of our CopyWithCallback() function is simply to copy data
from a source to a destination, while periodically invoking a callback
function to write progress messages to a log file, if one is defined
with the GIT_LFS_PROGRESS environment variable.  These progress log
entries include a field which represents the amount of an object's
data that has been transferred so far as a fraction whose denominator
is the expected total size of the object.

To populate the denominator of this field in the progress log
entries, the CallbackReader structure in our "tools" package is
initialized with a "totalSize" parameter, which will then be
passed to the provided callback function each time it is invoked.

Since the CopyWithCallback() function creates a CallbackReader
structure, it therefore requires both its own "totalSize" parameter
and a pointer to a callback function.  During a "clean" filter
operation, these parameters are passed down from the clean() function
to the GitFilter structure's Clean() method and then on to the
copyToTemp() method, which passes them to the CopyWithCallback()
function.  Note that the GitFilter structure's methods use the name
"fileSize" for the parameter which is ultimately used for the
CopyWithCallback() function's "totalSize" parameter.

In commit b0d669c05bdba7a30750a9238f5c4d5255dfcbab of PR #3902 we
added a check to the copyToTemp() method which sets the callback
function pointer to nil if the "fileSize" parameter is negative.

As noted above, we revised the clean() function in a prior commit
in this PR so that it no longer attempts to read the size of the
file identified by its "fileName" parameter.  Since both the
"git lfs clean" and "git lfs filter-process" commands pass a
negative value to the clean() function, that is the value which
all "clean" filter operations now pass to the copyToTemp() method.

As a result of our changes, the check for a negative "fileSize"
parameter value added to the copyToTemp() method in PR #3902 will
always succeed during "clean" filter operations.  This effectively
disables all progress log output from these operations even though
the clean() function explicitly creates a callback function for
this purpose using the CopyCallbackFile() method of the GitFilter
structure.

The "progress log: file counts" test that we renamed and
updated in the recent PR #6288 provides a demonstration of
this change in the behaviour of the Git LFS client.  In commit
4a4e524249cad77ce86cce3369172f9cfc80839d of that PR we added to
the test a check of the number of entries with the operation name
"clean" that appear in a progress log after a "git add" command.
Since our change to the clean() function in the prior commit in
this PR, that check now finds no matching entries in the progress
log, and so the test currently fails.

To resolve this problem, we now simply remove the conditional block
in the copyToTemp() method which sets the callback function pointer
to nil if the "fileSize" parameter is negative.

Note, though, that when we introduced that conditional block in
PR #3902, we did so with the intention that it would act as a
fallback protection against the problem reported in #3884.  Prior
to the changes from PR #3902, the "git hash-object" command would
fail when its --path option's argument matched a Git LFS filter
entry from a "gitattributes" file and the command's input was
smaller than the size of the read buffer the Git LFS client uses
when decoding pointers.

In commit b0d669c05bdba7a30750a9238f5c4d5255dfcbab of PR #3902 we
resolved this problem by ensuring that the Read() method of what
was then the "pktlineReader" structure in our "git" package properly
handled subsequent calls after reading a non-zero number of bytes
and also encountering the EOF condition.  (Note that we later moved
this structure into our separate "git-lfs/pktline" repository.)

In the same commit we also attempted to protect against the issue
from #3884 with a separate change to the copyToTemp() method.
Specifically, we added the conditional block which sets the callback
function pointer to nil if the "fileSize" parameter is negative,
with the understanding that this would prevent the method from
writing a file to the working tree when the Git LFS client was
invoked by a "git hash-object" command.  As we explained in the
commit description, this change was meant to ensure that we
"don't process the callback to write the file in this case,
since we don't actually want to write into the working tree".

However, we apparently made this change under a misapprehension.
The callback function provided to the copyToTemp() method is not
responsible for writing to a file in the working tree.  In fact,
during a "clean" filter operation, no file will be written to the
working tree at all, including by Git.  (This stands in contrast
to a "smudge" filter operation, when Git does write a file to the
working tree.)

What does occur during a "clean" filter operation, though, is
that the copyToTemp() method creates a temporary file in the
local Git LFS storage directories and writes to the file using
the CopyWithCallback() function from our "tools" package.
This occurs regardless of whether a callback function is defined
or not.  The CopyWithCallback() function uses the Copy() function
from the Go standard library's "io" package to perform the actual
transfer of data.  The clean() function in our "commands" package
then moves the temporary file into a final location in the object
storage directory hierarchy within the local Git LFS storage
directory.  All of these steps are performed even if the "clean"
filter operation was invoked via a "git hash-object" command and
not a "git add" command.

As described further above, the only purpose of the callback function
that the CopyCallbackFile() method of the GitFilter structure creates
is to write messages to a progress log file, if one is specified with
the GIT_LFS_PROGRESS environment variable.  So the conditional block
that we added to the copyToTemp() method in commit
b0d669c05bdba7a30750a9238f5c4d5255dfcbab only causes the progress
log to be disabled under some conditions, not to prevent a Git LFS
object file from being created.

Thus we can remove this conditional block entirely without concern
that we might introduce a regression and a re-introduction of the
problem reported in #3884.

However, absent some of our other changes from PR #6288, we would
cause a regression of another form, by re-introducing the issue
that was reported and resolved in PR #5504.  That PR revised the
callback function created by the CopyCallbackFile() method so that
it would only write a message to the progress log if at least 200
milliseconds had elapsed since the previous message was written, or
if all the data for an object had been read and so the current
message would be the final entry for a given object.  We made this
change specifically to avoid excessive numbers of writes to the
progress log during "clean" filter operations.

In order to ensure that a final progress log message was written for
an object when its data had been fully read, the callback function
would check whether its "written" parameter was equal to or larger
than its "total" parameter, and if so, unconditionally write a
message even if less than 200 milliseconds had elapsed since the
previous message was written.  This logic depends on the "total"
parameter's value never being negative, though, because if it is
then the "written" parameter will always be larger and so every
message will be written, regardless of the amount of time that
has elapsed since the previous message was written.

On their own, our changes to the clean() function in our prior
commit in this PR would result in the callback function writing
every message to the progress log during "clean" filter operations,
which would re-introduce the problem that we tried to resolve in
PR #5504.

Fortunately, though, we have recently addressed this issue in
PR #6288 by refactoring both the callback function created by the
CopyCallbackFile() method and the internal implementation of the
Read() method of the CallbackReader structure.  The callback
function now avoids writing messages if insufficient time has
elapsed even when the "total" parameter's value is negative.

As well, when the callback function receives a negative value in
its "total" parameter, it represents the unknown total as a series
of question mark characters in the progress log entry, rather than
as a literal negative integer.

Further, the internal implementation of the CallbackReader
structure's Read() method will now change the value of the "total"
parameter to equal that of the "written" parameter when it invokes
the callback function for the final time for a given object.
This ensures that the final entry for an object whose size was
initially unknown will not contain question mark characters
(or a negative integer) but will instead have equal values in
the numerator and denominator of the field which represents the
amount of the object's data that has been processed.

To verify that these changes in PR #6288 work in combination with
our changes from this PR to generate an entry in the progress log
during a "clean" filter operation, we now add a "progress log:
unknown initial size (clean)" test to our "t/t-progress.sh" shell
script.  We model this test on the "progress log: unknown initial
size (smudge)" and "progress log: unknown initial size (fetch/pull)"
tests that we added in commit 47d1e253a3ecc9546acf2546c7db634fe2c502bf
of PR #6288.

Our new test performs a "git add" command and confirms that although
the "git lfs filter-process" command no longer attempts to determine
the size of an object by calling the Stat() function on the
corresponding file in the working tree, a final progress log entry
will be written where the total size of the object is correctly
reported.

chrisd8088 commented on t/cmd/lfstest-badlocalfile.go:
```diff
@@ -0,0 +1,72 @@
+						}
+						if err = f.Truncate(0); err != nil {
+							panic(err)
+						}
```
Here I think we can slightly simplify the action of this utility by just forcing a new file to be created, which by definition will be empty:

```suggestion
						f, err := os.OpenFile(pathname, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o644)
						if err != nil {
							panic(err)
						}
```

chrisd8088 commented on t/t-filter-process.sh:
```diff
@@ -317,3 +317,33 @@ begin_test "filter process: git archive does not invoke SSH"
+  git add .gitattributes
+  git commit -m "initial commit"
+
+  dd if=/dev/zero of=large.dat bs=65537 count=1
```
Rather than create a file of a fixed size, we can now take advantage of the `lfstest-getlimit` [helper](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/t/cmd/lfstest-getlimit.go) we added in PR #6285 to ensure that we create a file large enough to exceed the [size](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/lfs/scanner.go#L10-L12) of the internal buffer we use when [checking](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/lfs/gitfilter_clean.go#L79-L86) whether a file already contains a valid Git LFS pointer.

In PR #6285 we updated some [other tests](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/t/t-filter-process.sh#L193-L255) in this script to use the `lfstest-getlimit` helper, so following their current design, we can do something like the following here:

```suggestion
  # Test with a file size larger than the size of the read buffer used when
  # decoding pointers.  See the secondary issue reported in
  # https://github.com/git-lfs/git-lfs/issues/4974#issuecomment-2717907543
  # and https://github.com/git-lfs/git-lfs/pull/6011.
  max_pointer_size="$(lfstest-getlimit --max-pointer-size)"
  contents_size=$((max_pointer_size * 2))
  head -c "$contents_size" /dev/zero >large.dat
```

chrisd8088 commented on t/t-filter-process.sh:
```diff
@@ -317,3 +317,33 @@ begin_test "filter process: git archive does not invoke SSH"
+    -c "filter.lfs.clean=false"\
+    -c "filter.lfs.smudge=false" \
+    -c "filter.lfs.required=true" \
+    add large.dat
```
If we capture the trace log output from the `git add` command, which includes the trace log output from the `git lfs filter-process` command as well, we can check to make sure the `unknown command` error message no longer appears:

```suggestion
    add large.dat 2>&1 | tee filter.log
  [ 0 -eq "${PIPESTATUS[0]}" ]

  [ 0 -eq "$(grep -c "unknown command" filter.log)" ]
```

chrisd8088 commented on commands/command_clean.go:
We should rewrite this comment to reflect the revised action of the function:

```diff
-// If fileSize is given as a non-negative (>= 0) integer, that value is used
-// with preference to os.Stat(fileName).Size(). If it is given as negative, the
-// value from the `stat(1)` call will be used instead.
+// If the total size of the object is unknown until it is read, fileSize
+// should be a negative integer.

```

chrisd8088 reviewed (commented):
Thank you again for this PR, and my apologies for the amount of time it's taken for us to get it merged!

I've already written commits which address all of my notes and suggestions from this belated review, along with some other minor cleanup issues, and I'll push those up to this PR once we've merged PR #6288 into `main` and I can cleanly pull its changes into this PR's branch.
 
In other words, there should be no need for you to try to address any of my comments; they are just here for future reference.

---

The key hurdle, as you [noted](https://github.com/git-lfs/git-lfs/pull/6011/#issuecomment-2761229659), is that by ignoring the file in the working tree which (usually!) corresponds to the object we're processing through our `clean` filter, that means we'll always pass a `-1` down to the `copyToTemp()` method of our `GitFilter` structure, and so if we make no other changes, that method will always [unset](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/lfs/gitfilter_clean.go#L73-L75) the callback function pointer we explicitly [create](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/commands/command_clean.go#L38) for `clean` filter operations.  As a result, we'd effectively disable progress logging during these operations, even when it should be enabled because the `GIT_LFS_PROGRESS` environment variable specifies a log file path.

We added the conditional block in the `copyToTemp()` method back in PR #3902 with intent that it would cause our `clean` filter operations to skip creating an object file in local storage when there was no need for it, for instance, if the invoking command was `git hash-object` rather than `git add`.  However, I believe we were under the misapprehension that the callback function is responsible for creating an object file, when its only purpose is to write an entry in the progress log (if a log file has been specified with the `GIT_LFS_PROGRESS` environment variable).

So we can just remove this conditional block and allow the `-1` values to be passed down to the callback function.  However, doing so exposes another problem, which is that in PR #5504 we revised the anonymous callback function [created](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/lfs/util.go#L56-L67) by the `CopyCallbackFile()` method of the `GitFilter` structure so that it throttles its writes to the progress log, and we explicitly did so to reduce the progress log I/O overhead during `clean` filter operations.

To throttle its writes, the callback function skips writing an entry to the log if less than 200 ms have elapsed since the previous write for the same object and the `written` parameter is less than the `total` parameter.  That logic ensures a final entry is written regardless of the amount of time elapsed since the previous write, but only if the `total` parameter is valid.  If it's `-1`, though, the throttling logic will be defeated and every callback will result in a write, which would re-introduce the problem PR #5504 was meant to solve.

That problem is what led to the work in PR #6288, which in part ensures the throttling works in all cases, even when `total` is negative.  That PR's changes also mean we won't write a literal `-1` into the progress log anymore (something we already do, per my notes in https://github.com/git-lfs/git-lfs/pull/6011#issuecomment-3866531617), and will use a `????`-style field instead.

There's also a lot of other cleanup that went into that PR, to deal with the fact that for historical reasons we actually have two functions which format [identical](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/lfs/util.go#L59) [messages](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/tq/meter.go#L278) for the progress log, and (separately) [two](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/tools/copycallback.go#L70-L75) [structures](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/tools/copycallback.go#L11-L16) which wrap `io.Reader` and `io.ReadSeekCloser`, respectively, and these are used in three different combinations to generate entries in the log.

---

We have two other commands besides `git lfs clean` and `git lfs filter-process` which also call the `clean()` function, the `git lfs migrate import` command and the `git lfs merge-driver` command.  In the case of the `git lfs migrate import` command, we actually know the size of a given object for certain in the [two](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/commands/command_migrate_import.go#L412) [instances](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/commands/command_migrate_import.go#L176) where we call the `clean()` function, because we're reading a blob from repository's Git history.

However, in the case of the `git lfs merge-driver` command, we [pipe](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/commands/command_merge_driver.go#L39) the output of another command into a temporary file (the one identified by the `%D` specifier), and then [make](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/commands/command_merge_driver.go#L76-L79) an `os.Stat()` call just to find the size of that file before [passing](https://github.com/git-lfs/git-lfs/blob/ac3a0f402316f8b1e29e6348f4a630adb6512539/commands/command_merge_driver.go#L88) it to the `clean()` function.

Since it's at least theoretically possible for the temporary file to change between the time we read its size and the time we open it, we should just skip the `os.Stat()` call and pass `-1` to the `clean()` command in its `fileSize` parameter.

chrisd8088 committed:
Merge branch 'main' into cynix-fix-clean
